Piwik.Tracker.Tests/PiwikTrackerTests.cs

```diff
@@ -1,62 +1,591 @@
+            var actual = _sut.GetCustomVariable(variableId, variableScope);
+            //Assert
+            Assert.That(actual, Is.Null);
+        }
 
         [Test]
-        [TestCase(Scopes.Page, null, null, null)]
         [TestCase(Scopes.Page, 2, null, null)]
         [TestCase(Scopes.Page, 2, "myPageVar", "myPageVarValue")]
-        [TestCase(Scopes.Event, null, null, null)]
         [TestCase(Scopes.Event, 3, null, null)]
         [TestCase(Scopes.Event, 3, "myEventVar", "myEventVarValue")]
-        [TestCase(Scopes.Visit, null, null, null)]
         [TestCase(Scopes.Visit, 4, null, null)]
         [TestCase(Scopes.Visit, 4, "myVisitVar", "myVisitVarValue")]
-        public void GetCustomVariable_Test(Scopes variableScope, int? variableId, string variableName, string variableValue)
+        public void GetCustomVariable_WhenVariableIsSet_ReturnsCorrectVariable(Scopes variableScope, int variableId, string variableName, string variableValue)
+        {
+            //Arrange
+            _sut.SetCustomVariable(variableId, variableName, variableValue, variableScope);
+
+            //Act
+            var actual = _sut.GetCustomVariable(variableId, variableScope);
+            //Assert
+            Assert.That(actual.Name, Is.EqualTo(variableName));
+            Assert.That(actual.Value, Is.EqualTo(variableValue));
+        }
+
+        [Test]
+        public void GetCustomVariable_WhenInvalidScopeArgument_Throws()
         {
             //Arrange
-            var sut = new PiwikTracker(SiteId, PiwikBaseUrl);
-            if (variableId != null)
+            //Act & Assert
+            Assert.Throws<ArgumentException>(() => _sut.GetCustomVariable(1, (Scopes)1234));
+        }
+
+        [Test]
+        public void Ctor_WhenNoUrlProvided_Throws()
+        {
+            Assert.Throws<ArgumentException>(() => new PiwikTracker(0, null));
+        }
+
+        [Test]
+        [TestCase("testCharset", "&cs=testCharset")]
+        [TestCase("utf16", "&cs=utf16")]
+        [TestCase("", null)]
+        public void SetPageCharset_WhenSpecified_IsAddedToRequest(string charset, string expected)
+        {
+            // Arrange, Act
+            _sut.SetPageCharset(charset);
+            // Assert
+            var actual = _sut.GetRequest(SiteId);
+            if (string.IsNullOrEmpty(expected))
             {
-                sut.SetCustomVariable(variableId.Value, variableName, variableValue, variableScope);
+                Assert.That(actual, Does.Not.Contain("&cs="));
             }
-            //Act
+            else
+            {
+                Assert.That(actual, Does.Contain(expected));
+            }
+        }
 
-            var actual = sut.GetCustomVariable(variableId ?? 99, variableScope);
-            //Assert
-            if (variableId != null)
+        [Test]
+        [TestCase("http://myurl.com/index.html")]
+        [TestCase("")]
+        public void SetUrl_WhenSpecified_IsAddedToRequest(string pageUrl)
```

Comment:
Hi Julien,

I totally agree with the things you said about a good testing approach.
Maybe I did not become a big fan of this mock server approach since it did not help me to reconstruct the behavior we experienced with the not disposed request. For me, this mocking server approach is a little bit of a patch for bad software design. I also thought of making the piwik tracker code more testable but I did not want to introduce further complexity / ideas in this unit test PR.
Writing the tests took more time than expected because of the huge amount of methods from piwik tracker and I am not sure whether we would have less code / same test coverage when I would have made more tests with the mocked server.

Btw: I was surprised that I found as much duplicated code in my tests, this shows that I was not very focused at the end...